### PR TITLE
sql: collect contention stats for sql stats when tracing is enabled

### LINKED ISSUE #78675: sql: contention duration not recorded properly in SQL Stats
Discovered while working on #73566.

When a contention event is experienced, the contention event is recorded in the contention event registry. However, the same event is not reflected in the SQL Stats. 

Observed using the following injected code:

`conn_executor_exec.go` 

```golang
@@ -2188,8 +2188,27 @@ func (ex *connExecutor) recordTransactionFinish(
        }
        ex.metrics.EngineMetrics.SQLTxnLatency.RecordValue(txnTime.Nanoseconds())

+       s := "internal"
+       if ex.executorType == executorTypeExec {
+               s = "external"
+       }
+
        if contentionDuration := ex.extraTxnState.accumulatedStats.ContentionTime.Nanoseconds(); contentionDuration > 0 {
                ex.metrics.EngineMetrics.SQLContendedTxns.Inc(1)
+               if ex.executorType == executorTypeExec {
+                       fmt.Printf("[DEBUG] (%s)(%s): recording contention: %d (%s) | n%s\n",
+                               s,
+                               ev.txnID.String(),
+                               ex.metrics.EngineMetrics.SQLContendedTxns.Count(),
+                               ex.extraTxnState.accumulatedStats.ContentionTime.String(),
+                               ex.server.cfg.NodeID.SQLInstanceID().String())
+               }
+       } else {
+               if ex.executorType == executorTypeExec {
+                       fmt.Printf("[DEBUG]: (%s|n%s)(%s) no contention observed: %s\n",
+                               s, ex.server.cfg.NodeID.SQLInstanceID().String(),
+                               ev.txnID.String(), ex.extraTxnState.accumulatedStats.ContentionTime.String())
+               }
        }

        ex.txnIDCacheWriter.Record(contentionpb.ResolvedTxnID{
```

In `distsql_running.go` 

```golang
--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -1004,6 +1004,7 @@ func (r *DistSQLReceiver) pushMeta(meta *execinfrapb.ProducerMetadata) execinfra
                                        contentionEvent.WaitingTxnID = r.txn.ID()
                                }
                                r.contentionRegistry.AddContentionEvent(contentionEvent)
+                               fmt.Printf("[DEBUG]: contention event for txnID: %s, %s\n", r.txn.ID(), contentionEvent.BlockingEvent.Duration.String())
                        })
                }
        }
```

The following log entry shows the discrepancy:

```
[DEBUG]: contention event for txnID: 789b206f-b507-4f25-ad24-c11a62f664db, 6.571134ms
[DEBUG]: (external|n2)(789b206f-b507-4f25-ad24-c11a62f664db) no contention observed: 0s
``` 

Same discrepancy can be observed in `crdb_internal.transaction_contention_events` table and the SQL Stats tables. 

Jira issue: CRDB-14212

### PR DESCRIPTION
**Pull request by Azhng**
Resolves #78675

Previously, contention stats was not collected for SQL Stats even when
tracing was enabled. This was caused by two bugs:
1. instrumentationHelper skipping analyzing traces when tracing is enabled
2. transaction statistics ignore traces when tracing is turned on at the
higher level.

This commit ensures that contention stats is collected when tracing is
turned on.

Release note (bug fix): Contention statistics are now being collected for
SQL Stats when tracing is enabled.

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Nice job tracking this down!

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-:-MzXswU56wDi22oD3GNj:b-8botvf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/sql/conn_executor_exec.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXrOGN2KcA0yQoT5Sl:-MzXrOGN2KcA0yQoT5Sm:bg8sjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L473)):*
> ```Go
> 				ex.statsCollector,
> 				&ex.extraTxnState.accumulatedStats,
> 				ex.extraTxnState.shouldCollectTxnExecutionStats || ih.collectExecStats,
> ```

Hm, this seems suspicious. This argument to `Finish` determines whether we accumulate `txnStats`, and the fact that we happened to be collecting exec stats for a statement shouldn't matter. Imagine the following scenario:
- start a new txn. We decide to not sample txn stats for it, so `shouldCollectTxnExecutionStats == false`
- we enable verbose tracing
- run some statements (1), for all of them exec stats are collected
- disable verbose tracing
- run some more statements (2), for all of these exec stats are not collected.

If we were accumulate txn stats because we happened to be collecting some exec stats, our txn stats would be partial - containing the info only from (1).

What do you think? Does having a partial info in txn stats better than no info? I feel like having the partial info might be misleading / confusing, so I'd favor no info.
___
*[pkg/sql/conn_executor_exec.go, line 2221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXp_VJ9zXk_fmNJxGr:-MzXp_VKA-AFWpq_gpFA:bgdgpru) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L2221)):*
> ```Go
> 	commitLat := ex.phaseTimes.GetCommitLatency()
> 
> 	collectedExecStats :=
> ```

nit: I don't think this is needed - `ex.planner.instrumentation.collectExecStats` will always be `true` when `shouldCollectTxnExecutionStats` is `true`, so we can just use `ex.planner.instrumentation.collectExecStats`.
___
*[pkg/sql/instrumentation.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXqDUo9ycgbTsHkwuf:-MzXqDUo9ycgbTsHkwug:b54xs0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/instrumentation.go#L199)):*
> ```Go
> 			// no extra work is needed by the instrumentationHelper.
> 			ih.collectExecStats = true
> 			// We still, however, want to finish the instrumentationHelper in
> ```

nit: this comment needs an adjustment. I'd reword to something like:
```
We always want to finish the instrumentationHelper in order
to record the execution statistics. Note that we capture the
span in order to fetch the trace from it, but the span won't be
finished.
```
___
*[pkg/sql/instrumentation.go, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXsMu3BKVpDeSIFwsB:-MzXsMu3BKVpDeSIFwsC:b2vu950) ([raw file](https://github.com/cockroachdb/cockroach/blob/3272fcec17cbc6027f987fbc630d0459307f6572/pkg/sql/instrumentation.go#L317)):*
> ```Go
> 			}
> 		}
> 		if collectTxnExecStats || ih.implicitTxn {
> ```

If my comment above makes sense and we revert to the original value passed for `collectTxnExecStats`, then this line should also be reverted.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @kevin-v-ngo)

___
*[pkg/sql/conn_executor_exec.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXrOGN2KcA0yQoT5Sl:-N--6k9H8TFzlUP3S3j_:bu9zs20) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L473)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, this seems suspicious. This argument to `Finish` determines whether we accumulate `txnStats`, and the fact that we happened to be collecting exec stats for a statement shouldn't matter. Imagine the following scenario:
- start a new txn. We decide to not sample txn stats for it, so `shouldCollectTxnExecutionStats == false`
- we enable verbose tracing
- run some statements (1), for all of them exec stats are collected
- disable verbose tracing
- run some more statements (2), for all of these exec stats are not collected.

If we were accumulate txn stats because we happened to be collecting some exec stats, our txn stats would be partial - containing the info only from (1).

What do you think? Does having a partial info in txn stats better than no info? I feel like having the partial info might be misleading / confusing, so I'd favor no info.
</blockquote></details>

Hmm, good question, I feel this will be confusing either way:
* Suppose we have partial txn execution stats info, then the execution stats might not realistically reflect what happened during the transaction execution, e.g. we have

```sql
BEGIN;
<stmt1> -- spend 1 seconds in contention, contention collected via SET tracing=on
<stmt2> -- spend 1 hour in contention, contention not collected via SET tracing=off
COMMIT;
```
Then the transaction stats will only show 1 second of contention even though user would observe this transaction hanged for an hour. 

* Suppose we have no txn execution stats info, then user would observe a contradiction:
  * statement stats shows there's a contention during the execution of the statement
  * transaction stats shows there's no contention during the execution of the transaction that the statement was a part of

cc @kevin-v-ngo  thoughts from the product perspective on this?

**Comment by kevin-v-ngo** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @kevin-v-ngo)
> 
> _[pkg/sql/conn_executor_exec.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXrOGN2KcA0yQoT5Sl:-N--6k9H8TFzlUP3S3j_:bu9zs20) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L473)):_
> 
> Previously, yuzefovich (Yahor Yuzefovich) wrote…
> Hmm, good question, I feel this will be confusing either way:
> 
> * Suppose we have partial txn execution stats info, then the execution stats might not realistically reflect what happened during the transaction execution, e.g. we have
> 
> ```sql
> BEGIN;
> <stmt1> -- spend 1 seconds in contention, contention collected via SET tracing=on
> <stmt2> -- spend 1 hour in contention, contention not collected via SET tracing=off
> COMMIT;
> ```
> 
> Then the transaction stats will only show 1 second of contention even though user would observe this transaction hanged for an hour.
> 
> * Suppose we have no txn execution stats info, then user would observe a contradiction:
>   
>   * statement stats shows there's a contention during the execution of the statement
>   * transaction stats shows there's no contention during the execution of the transaction that the statement was a part of
> 
> cc @kevin-v-ngo thoughts from the product perspective on this?

Thanks Archer and Yahor! Good catch. It’s interesting how we haven’t seen users report this issue before. I guess not too many use EXPLAIN ANALYZE DEBUG or manual SESSION tracing to notice lack of contention stats. For ambiguous functional scenarios like this, I normally favor the behavior that is the easiest explainable to users. 

In this case, the easy explanation seems to be that we only capture (contention) stats for when tracing is enabled. This tracing must be _turned on manually by the user correct_? If so, they should be able to understand that the partial contention stats (your example) was only for requests that ran when they manually turned on tracing. 

I'd be in favor of no info if it's unclear to the user that tracing was turned on and off but this is manually configured. Let me know if i missed anything.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXrOGN2KcA0yQoT5Sl:-N-3SPFr9fEWCorlPkhT:b-jsmewk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L473)):*
<details><summary><i>Previously, kevin-v-ngo wrote…</i></summary><blockquote>

Thanks Archer and Yahor! Good catch. It’s interesting how we haven’t seen users report this issue before. I guess not too many use EXPLAIN ANALYZE DEBUG or manual SESSION tracing to notice lack of contention stats. For ambiguous functional scenarios like this, I normally favor the behavior that is the easiest explainable to users. 

In this case, the easy explanation seems to be that we only capture (contention) stats for when tracing is enabled. This tracing must be _turned on manually by the user correct_? If so, they should be able to understand that the partial contention stats (your example) was only for requests that ran when they manually turned on tracing. 

I'd be in favor of no info if it's unclear to the user that tracing was turned on and off but this is manually configured. Let me know if i missed anything.

</blockquote></details>

Thanks for the product perspective here @kevin-v-ngo! Thinking about it more, I realized that statement bundle can be a bit problematic here. This is the case that /* kind of */ auto enables tracing for a statements. 

In this case, if the statement matches requested statement fingerprint, the tracing is automatically enabled. 
Any user with `VIEWACTIVITYREDACTED` permission can request for statement bundle. 

I will go ahead to show no info for the txn stats.

**Comment by kevin-v-ngo** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @yuzefovich)
> 
> _[pkg/sql/conn_executor_exec.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXrOGN2KcA0yQoT5Sl:-N-3SPFr9fEWCorlPkhT:b-jsmewk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L473)):_
> 
> Previously, kevin-v-ngo wrote…
> Thanks for the product perspective here @kevin-v-ngo! Thinking about it more, I realized that statement bundle can be a bit problematic here. This is the case that /* kind of */ auto enables tracing for a statements.
> 
> In this case, if the statement matches requested statement fingerprint, the tracing is automatically enabled. Any user with `VIEWACTIVITYREDACTED` permission can request for statement bundle.
> 
> I will go ahead to show no info for the txn stats.

Doesn't the user specify turning on tracing for the specific statement by running EXPLAIN ANALYZE DEBUG or via activate conditional statements? I guess in this case, there are no transaction level stats (just statement) so we have the contention metrics already.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXrOGN2KcA0yQoT5Sl:-N-4PzhR5vPKG9VBribn:b-3vap1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L473)):*
<details><summary><i>Previously, kevin-v-ngo wrote…</i></summary><blockquote>

Doesn't the user specify turning on tracing for the specific statement by running EXPLAIN ANALYZE DEBUG or via activate conditional statements? I guess in this case, there are no transaction level stats (just statement) so we have the contention metrics already.
</blockquote></details>

Chatted offline, will go ahead to allow for partial contention stats to be collected for transaction stats.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 2221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXp_VJ9zXk_fmNJxGr:-N-56dPeCLEmIQ9sBGod:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/conn_executor_exec.go#L2221)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I don't think this is needed - `ex.planner.instrumentation.collectExecStats` will always be `true` when `shouldCollectTxnExecutionStats` is `true`, so we can just use `ex.planner.instrumentation.collectExecStats`.
</blockquote></details>

Done.
___
*[pkg/sql/instrumentation.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXqDUo9ycgbTsHkwuf:-N-58xbN7KaDwi2ZaOmc:bwf8y2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/instrumentation.go#L199)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this comment needs an adjustment. I'd reword to something like:
```
We always want to finish the instrumentationHelper in order
to record the execution statistics. Note that we capture the
span in order to fetch the trace from it, but the span won't be
finished.
```
</blockquote></details>

Done.

Quick question here, does "finishing a span" here means either:
* the tracing span can no longer collect more data? Or
* the memory allocated for the span is freed? 
___
*[pkg/sql/instrumentation.go, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXsMu3BKVpDeSIFwsB:-N-5AxKQBHf8low-3ZEk:blfvsga) ([raw file](https://github.com/cockroachdb/cockroach/blob/3272fcec17cbc6027f987fbc630d0459307f6572/pkg/sql/instrumentation.go#L317)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

If my comment above makes sense and we revert to the original value passed for `collectTxnExecStats`, then this line should also be reverted.
</blockquote></details>

Renamed `collectTxnExecStats` to `collectExecStats`. Reverted the change here, I kinda forgot why I made that change in the first place and seems like the code is fine with the additional `ih.implicitTxn` variable in the `if` condition

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-:-N-5ps-PDH11o22uCHPZ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/sql/instrumentation.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79175#-MzXqDUo9ycgbTsHkwuf:-N-5ozpC-LVa_fk7WOhb:b-ph3h9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7ee8c6d074210b55023074674835980689de4e0/pkg/sql/instrumentation.go#L199)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Done.

Quick question here, does "finishing a span" here means either:
* the tracing span can no longer collect more data? Or
* the memory allocated for the span is freed? 
</blockquote></details>

I think both. Once finished, the span object might be put back into `sync.Pool` and reused.

**Comment by Azhng** [comment]
blathers backport release-22.1
